core/blockchain.go

```diff
@@ -1736,3 +1743,47 @@ func (bc *BlockChain) SubscribeLogsEvent(ch chan<- []*types.Log) event.Subscript
 func (bc *BlockChain) SubscribeBlockProcessingEvent(ch chan<- bool) event.Subscription {
 	return bc.scope.Track(bc.blockProcFeed.Subscribe(ch))
 }
+
+// GetAncestorHash return the hash of ancestor at the given number
+func (bc *BlockChain) GetAncestorHash(ref *types.Header, target uint64) common.Hash {
+	number := ref.Number.Uint64() - 1
+	hash := ref.ParentHash
+	if target == number {
+		return hash
+	}
+	if target > number {
+		// Should never happen
+		log.Error("Ancestor number must be <= descendant", "ancestor target", target, "descendant", number)
+		return common.Hash{}
+	}
+	// Iterate the chain until we hit the target or we hit a ancestor
+	// within the storage
+	for !bc.hashHistory.Is(number, hash) {
+		header := bc.GetHeader(hash, number)
+		if header == nil {
+			return common.Hash{}
+		}
+		number = header.Number.Uint64() - 1
+		hash = header.ParentHash
+		if number == target {
+			return hash
+		}
+	}
+	// The hash storage has the right ancestor chain
+	if h, found := bc.hashHistory.Get(target); found {
```

Comment:
Will execution ever reach here? Further, it seems like `h`will be found and returned in the `for`-loop above when `number == target`.

Reply:
Yes, we might be executing on a sidechain. The `hashstorage` will store the canon chain. So if we have
`[..., b1, b2, b3, b4, b5, b6]` and we're executing on a sidechain which is 
`[..., b1, b2, c3, c4, c5, c6]`, and the requested number is `b1` from `c6`. Then the loop above will slow-iterate through `c5,c3,c2` and discover that `b2` actually is part of the `hashHistory`. And then it can do a direct lookup of `b1`.

Reply:
Thanks for clarifying. I confirm that's what I now see as well.